packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -0,0 +1,413 @@
+              response.statusMessage;
+            span.setStatus(Utils.parseResponseStatus(response.statusCode));
+          }
+
+          span.setAttributes(attributes);
+
+          if (this.options.applyCustomAttributesOnSpan) {
+            this.options.applyCustomAttributesOnSpan(span, request, response);
+          }
+
+          span.end();
+        });
+        Utils.setSpanOnError(span, response);
+      });
+
+      Utils.setSpanOnError(span, request);
+
+      this._logger.debug('makeRequestTrace return request');
+      return request;
+    };
+  }
+
+  private _incomingRequestFunction(
+    original: (event: string, ...args: unknown[]) => boolean
+  ) {
+    const plugin = this;
+    return function incomingRequest(
+      this: {},
+      event: string,
+      ...args: unknown[]
+    ): boolean {
+      // Only traces request events
+      if (event !== 'request') {
+        return original.apply(this, [event, ...args]);
+      }
+
+      const request = args[0] as IncomingMessage;
+      const response = args[1] as ServerResponse;
+      const pathname = request.url
+        ? url.parse(request.url).pathname || '/'
+        : '/';
+      const method = request.method || 'GET';
+
+      plugin._logger.debug('%s plugin incomingRequest', plugin.moduleName);
+
+      if (
+        Utils.isIgnored(pathname, request, plugin.options.ignoreIncomingPaths)
+      ) {
+        return original.apply(this, [event, ...args]);
+      }
+
+      const propagation = plugin._tracer.getHttpTextFormat();
+      const headers = request.headers;
+      const spanOptions: SpanOptions = {
+        kind: SpanKind.SERVER,
+      };
+
+      const spanContext = propagation.extract(Format.HTTP, headers);
+      if (spanContext && isValid(spanContext)) {
+        spanOptions.parent = spanContext;
+      }
+
+      const span = plugin._startHttpSpan(`${method} ${pathname}`, spanOptions);
+
+      return plugin._tracer.withSpan(span, () => {
+        plugin._tracer.bind(request);
+        plugin._tracer.bind(response);
+
+        // Wraps end (inspired by:
+        // https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/master/src/plugins/plugin-connect.ts#L75)
+        const originalEnd = response.end;
+        response.end = function(
+          this: ServerResponse,
+          ...args: ResponseEndArgs
+        ) {
+          response.end = originalEnd;
+          // Cannot pass args of type ResponseEndArgs,
+          // tslint complains "Expected 1-2 arguments, but got 1 or more.", it does not make sense to me
+          // tslint:disable-next-line:no-any
+          const returned = response.end.apply(this, arguments as any);
+          const requestUrl = request.url ? url.parse(request.url) : null;
+          const hostname = headers.host
+            ? headers.host.replace(/^(.*)(\:[0-9]{1,5})/, '$1')
+            : 'localhost';
+          const userAgent = (headers['user-agent'] ||
+            headers['User-Agent']) as string;
```

Comment:
headers for incoming messages are always lowercase according to https://nodejs.org/dist/latest-v12.x/docs/api/http.html#http_message_headers therefore the extra check for `User-Agent` should be not needed.